File: extract_info.R
# Load necessary libraries
library(httr)
library(jsonlite)
library(dplyr)
library(readr)

# Function to extract information using local LLM via Ollama
extract_info <- function(title, abstract) {
  prompt <- paste0(
    "Analyze the following paper title and abstract. Extract the following information if available:\n",
    "- Forest type (e.g., boreal, temperate, tropical, coniferous, deciduous, mixed)\n",
    "- Region (e.g., country, continent, specific location)\n",
    "- Scale (e.g., plot, stand, landscape, regional, global)\n",
    "- Statistical method (e.g., linear regression, random forest, SVM, neural network)\n",
    "- R² values (list all reported R² or coefficient of determination)\n",
    "- RMSE values (list all reported RMSE or root mean square error)\n\n",
    "Title: ", title, "\n",
    "Abstract: ", abstract, "\n\n",
    "Output in JSON format:\n",
    "{\n",
    "  \"forest_type\": \"string or list\",\n",
    "  \"region\": \"string\",\n",
    "  \"scale\": \"string\",\n",
    "  \"statistical_method\": \"string or list\",\n",
    "  \"r2_values\": [list of floats],\n",
    "  \"rmse_values\": [list of floats]\n",
    "}\n",
    "If not available, use null or empty list."
  )
  
  response <- POST(
    url = "http://localhost:11434/api/generate",
    body = toJSON(list(
      model = "llama3.2",  # or your chosen model
      prompt = prompt,
      stream = FALSE
    ), auto_unbox = TRUE),
    encode = "json"
  )
  
  if (status_code(response) == 200) {
    result <- content(response, "text")
    result_json <- fromJSON(result)
    return(result_json$response)
  } else {
    cat("Error:", status_code(response), "\n")
    return(NULL)
  }
}

# Read the TSV file
df <- read_tsv('/Users/jgoldman/git/sensor_scaling_literature_review/wos_search_refined_11-16-25.txt')

# Initialize list for results
results <- list()

# Process each paper
for (i in 1:nrow(df)) {
  title <- df$TI[i]
  abstract <- df$AB[i]
  doi <- df$DI[i]
  
  if (is.na(abstract) || abstract == "") next
  
  info_json <- extract_info(title, abstract)
  if (!is.null(info_json)) {
    info <- fromJSON(info_json)
    info$title <- title
    info$doi <- doi
    results <- append(results, list(info))
  }
  
  Sys.sleep(1)  # Rate limit
}

# Convert to DataFrame
output_df <- bind_rows(results)

# Save to CSV
write_csv(output_df, '/Users/jgoldman/git/sensor_scaling_literature_review/extracted_data_r.csv')

print("Extraction complete. Data saved to extracted_data_r.csv")

File: extract_info.py
import pandas as pd
import requests
import time
import re
import json

# Function to extract information using local LLM via Ollama
def extract_info(title, abstract):
    prompt = f"""
    Analyze the following paper title and abstract. Extract the following information if available:
    - Forest type (e.g., boreal, temperate, tropical, coniferous, deciduous, mixed)
    - Region (e.g., country, continent, specific location)
    - Scale (e.g., plot, stand, landscape, regional, global)
    - Statistical method (e.g., linear regression, random forest, SVM, neural network)
    - R² values (list all reported R² or coefficient of determination)
    - RMSE values (list all reported RMSE or root mean square error)

    Title: {title}
    Abstract: {abstract}

    Output in JSON format:
    {{
        "forest_type": "string or list",
        "region": "string",
        "scale": "string",
        "statistical_method": "string or list",
        "r2_values": [list of floats],
        "rmse_values": [list of floats]
    }}
    If not available, use null or empty list.
    """
    try:
        response = requests.post('http://localhost:11434/api/generate', json={
            "model": "llama3.2",  # or your chosen model
            "prompt": prompt,
            "stream": False
        })
        result = response.json()['response'].strip()
        # Assuming the model outputs JSON, parse it
        return json.loads(result)
    except Exception as e:
        print(f"Error: {e}")
        return None

# Read the TSV file
df = pd.read_csv('/Users/jgoldman/git/sensor_scaling_literature_review/wos_search_refined_11-16-25.txt', sep='\t', encoding='utf-8')

# Initialize list for results
results = []

# Process each paper
for idx, row in df.iterrows():
    title = row.get('TI', '')
    abstract = row.get('AB', '')
    doi = row.get('DI', '')
    
    if pd.isna(abstract) or abstract == '':
        continue  # Skip if no abstract
    
    info = extract_info(title, abstract)
    if info:
        info['title'] = title
        info['doi'] = doi
        results.append(info)
    
    time.sleep(1)  # Rate limit

# Convert to DataFrame
output_df = pd.DataFrame(results)

# Save to CSV
output_df.to_csv('/Users/jgoldman/git/sensor_scaling_literature_review/extracted_data.csv', index=False)

print("Extraction complete. Data saved to extracted_data.csv")